### PR TITLE
Add Gluon autolog capabilities for simple fitting.

### PR DESCRIPTION
**Pull request by cosmincatalin**
MXNet's Gluon API features a simple and direct for method on an estimator. This is similar to Keras.
This is the autologger for it.

## What changes are proposed in this pull request?

Basic functionality for tracking metrics and parameters of Gluon HybridBlocks based models. Autologging is included.

## How is this patch tested?

Added a gluon specific test as part of the pull request.

### Is this a user-facing change?

Autolog, save and load MXNet Gluon's HybridBlocks' based models.

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [x] API
- [ ] REST-API
- [x] Examples
- [x] Docs
- [ ] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [x] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [x] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/gluon.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+    >>> # Load persisted model as a Gluon model, make inferences against an NDArray
+    >>> model = mlflow.gluon.load_model("runs:/" + gluon_random_data_run.info.run_id + "/model")
+    >>> model(nd.array(np.random.rand(1000, 1, 32)))
+    """
+    local_model_path = _download_artifact_from_uri(artifact_uri=model_uri)
+
+    model_arch_path = os.path.join(local_model_path, "data", _MODEL_SAVE_PATH) + "-symbol.json"
+    model_params_path = os.path.join(local_model_path, "data", _MODEL_SAVE_PATH) + "-0000.params"
+    ctx = mxnet.cpu()
```
Comment (aarondav) [inline mlflow/gluon.py]: Not knowing much about gluon -- is it common to load using cpu instead of gpu? Should this be an option to the load_model?

File: tests/gluon_autolog/test_gluon_autolog.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import random
+import warnings
+
+import mxnet.ndarray as nd
```
Comment (aarondav) [inline tests/gluon_autolog/test_gluon_autolog.py]: I think we'll need to add mxnet to test requirements as well. CI is currently failing.

File: mlflow/gluon.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    symbol = sym.load(model_arch_path)
+    inputs = sym.var('data', dtype='float32')
+    net = gluon.SymbolBlock(symbol, inputs)
+    net.collect_params().load(model_params_path, ctx)
+    return net
+
+
+@experimental
+def save_model(gluon_model: HybridBlock, path: str, mlflow_model: Model = Model()):
```
Comment (aarondav) [inline mlflow/gluon.py]: MLflow currently still maintains compatibility with Python2 -- will these type annotations work there?

File: mlflow/gluon.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    >>> with net.name_scope():
+    >>> ...
+    >>> net.hybridize()
+    >>> net.collect_params().initialize()
+    >>> softmax_loss = SoftmaxCrossEntropyLoss()
+    >>> trainer = Trainer(net.collect_params())
+    >>> est = estimator.Estimator(net=net, loss=softmax_loss, metrics=Accuracy(), trainer=trainer)
+    >>> est.fit(train_data=train_data, epochs=100, val_data=validation_data)
+    >>> # Log metrics and log the model
+    >>> with mlflow.start_run() as run:
```
Comment (aarondav) [inline mlflow/gluon.py]: Probably this should wrap the fit call as well, right, to track at least the start/end times.

**Review by aarondav — COMMENTED** [review]
Overall, looks pretty good to me! I'd like to see if @dbczumar and @smurching could also take a look, as they're familiar with the models and autologging components specifically.

File: mlflow/gluon.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
+import os
```
Comment (dbczumar) [inline mlflow/gluon.py]: @cosmincatalin I noticed that this module doesn't produce MLflow Models with the `pyfunc` flavor. Is it possible to add a `pyfunc` representation so that the model is more broadly compatible with MLflow's supported model serving / deployment tools?

For reference, see the `mlflow.keras` module's pyfunc wrapper implementation: https://github.com/mlflow/mlflow/blob/7fd12b4decbee396fc7ecd0d828a99f288cc9397/mlflow/keras.py#L281-L333.

File: mlflow/gluon.py (hunk 1 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: Yes, it should be an option. I'll add that.

File: tests/gluon_autolog/test_gluon_autolog.py (hunk 2 above)
Comment (cosmincatalin) [reply tests/gluon_autolog/test_gluon_autolog.py]: I'll do that.

File: mlflow/gluon.py (hunk 3 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: I'm not an expert in Python so it's probably best if I remove them.

File: mlflow/gluon.py (hunk 5 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: Right, I'll try to do that.

File: mlflow/gluon.py (hunk 4 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: I'm not sure I get this part?

Comment (aarondav) [reply mlflow/gluon.py]: Just in the example, I mean it should probably be

```python
with mlflow.start_run() as run:
  est.fit(...)
  mlflow.log_mdoel
```

File: mlflow/gluon.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+                metric_name, metric_val = metric.get()
+                logs[metric_name] = metric_val
+            for metric in estimator.val_metrics:
+                metric_name, metric_val = metric.get()
+                logs[metric_name] = metric_val
+            try_mlflow_log(mlflow.log_metrics, logs, step=self.current_epoch)
+            self.current_epoch += 1
+
+        def train_end(self, estimator, *args, **kwargs):
+            try_mlflow_log(mlflow.log_param, "num_layers", len(estimator.net))
```
Comment (smurching) [inline mlflow/gluon.py]: As is done by https://github.com/mlflow/mlflow/pull/2055 for Keras autologging, maybe we should log these params at the beginning of training rather than at the end (in case training is interrupted / CTRL+C'd etc)

File: mlflow/gluon.py [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+            self.current_epoch = 0
+
+        def epoch_end(self, estimator, *args, **kwargs):
+            logs = {}
+            for metric in estimator.train_metrics:
+                metric_name, metric_val = metric.get()
+                logs[metric_name] = metric_val
+            for metric in estimator.val_metrics:
+                metric_name, metric_val = metric.get()
+                logs[metric_name] = metric_val
```
Comment (smurching) [inline mlflow/gluon.py]: I'm not super familiar with the gluon APIs, but is there any chance of the metric names colliding across training/validation? Looks like (from your test) the training metrics are prefixed with "train"

File: tests/gluon_autolog/test_gluon_autolog.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+            est.fit(data, epochs=5)
+
+    return client.get_run(run.info.run_id)
+
+
+@pytest.mark.large
+def test_gluon_autolog_logs_expected_data(gluon_random_data_run):
+    data = gluon_random_data_run.data
+    assert "train accuracy" in data.metrics
+    assert "train softmaxcrossentropyloss" in data.metrics
```
Comment (smurching) [inline tests/gluon_autolog/test_gluon_autolog.py]: Could we also add validation metrics to this test case?

File: mlflow/gluon.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+                try_mlflow_log(mlflow.log_param, "learning_rate",
+                               estimator.trainer.optimizer.lr)
+            if hasattr(estimator.trainer.optimizer, "epsilon"):
+                try_mlflow_log(mlflow.log_param, "epsilon",
+                               estimator.trainer.optimizer.epsilon)
+            if isinstance(estimator.net, HybridSequential):
+                try_mlflow_log(log_model, estimator.net, artifact_path="model")
+
+    @gorilla.patch(Estimator)
+    def fit(self, *args, **kwargs):
```
Comment (smurching) [inline mlflow/gluon.py]: As https://github.com/mlflow/mlflow/pull/2056/ does for Keras/TensorFlow, could we add logic to check if there's an active run here & if not (i.e. if we end up creating an active run within this `fit` call) add logic to end the run at the end of `fit`? It'd also be good to add a test case analogous to those in https://github.com/mlflow/mlflow/pull/2056/ for this behavior, thanks!

File: mlflow/gluon.py (hunk 6 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: I can probably do that. I'll look into it.

File: mlflow/gluon.py (hunk 7 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: That is correct, there should be no conflict as it stands right now.

File: mlflow/gluon.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+                try_mlflow_log(mlflow.log_param, "batches", estimator.max_batch)
+            try_mlflow_log(mlflow.log_param, "optimizer_name",
+                           type(estimator.trainer.optimizer).__name__)
+            if hasattr(estimator.trainer.optimizer, "lr"):
+                try_mlflow_log(mlflow.log_param, "learning_rate",
+                               estimator.trainer.optimizer.lr)
+            if hasattr(estimator.trainer.optimizer, "epsilon"):
+                try_mlflow_log(mlflow.log_param, "epsilon",
+                               estimator.trainer.optimizer.epsilon)
+            if isinstance(estimator.net, HybridSequential):
```
[the remark this answers is not in the document]
Comment (cosmincatalin) [reply mlflow/gluon.py]: Yes, only hybridizable models that have been hybridized can be exported. All other Gluon models are not exported right now. I can probably look into it in a future iteration. I hope that this is acceptable.

File: tests/gluon_autolog/test_gluon_autolog.py (hunk 8 above)
Comment (cosmincatalin) [reply tests/gluon_autolog/test_gluon_autolog.py]: I'll look into this. I expect to be able to add it.

File: mlflow/gluon.py (hunk 9 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: I will look into this.

File: mlflow/gluon.py (hunk 4 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: Right, got it now. I'll just update the documentation.

File: mlflow/gluon.py (hunk 5 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: The problem with adding a `pyfunc` for Gluon is that I cannot provided a needed additional argument, without modifying the `load_model` function in `pyfunc/__init__.py`, which I suspect you're not willing to easily accept.

Here is the code that would be needed in `gluon.py`:

```
class _GluonModelWrapper:
    def __init__(self, gluon_model):
        self.gluon_model = gluon_model

    def predict(self, ndarray):
        return self.gluon_model(ndarray)


def _load_pyfunc(path, ctx):
    """
    Load PyFunc implementation. Called by ``pyfunc.load_pyfunc``.

    :param path: Local filesystem path to the MLflow Model with the ``gluon`` flavor.
    :param ctx: Either CPU or GPU.
    """
    m = load_model(path, ctx)
    return _GluonModelWrapper(m)
```

You'll notice that I need to provide a context (CPU or GPU) when loading a model. To accomplish that I would need something similar to this in `pyfunc/__init__.py`:

```
def load_model(model_uri, suppress_warnings=True, **kwargs):
    ...
    return importlib.import_module(conf[MAIN])._load_pyfunc(data_path, **kwargs)
```

Otherwise, loading a Gluon model works fine like this:

```
model_loaded = mlflow.gluon.load_model(model_path, ctx.cpu())
```

Let me know your thought, @dbczumar

Comment (dbczumar) [reply mlflow/gluon.py]: @cosmincatalin Is it possible to have the `pyfunc` representation create and use a CPU context by default? This would remove the need for another argument to `_load_pyfunc`. 

Longer term, I think it would probably make sense to expand the `_load_pyfunc` interface to accept optional args/kwargs.

Comment (cosmincatalin) [reply mlflow/gluon.py]: I can do that, no problem.

**Comment by cosmincatalin** [comment]
All comments have been resolved. I can still see that Travis fails, but it does not seem to be related to what I have been doing. Can anyone help with that?

File: mlflow/gluon.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    # a specific epoch's paramaters, and is there only for display purposes.
+    gluon_model.export(os.path.join(data_path, _MODEL_SAVE_PATH))
+    with open(os.path.join(path, "architecture.txt"), "w") as fp:
+        fp.write(str(gluon_model))
+    pyfunc.add_to_model(mlflow_model, loader_module="mlflow.gluon")
+    mlflow_model.save(os.path.join(path, "MLmodel"))
+
+
+@experimental
+def log_model(gluon_model, artifact_path):
```
Comment (dbczumar) [inline mlflow/gluon.py]: @cosmincatalin Can you add a `conda_env` parameter to the `save_model` method?

File: mlflow/gluon.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+    Load PyFunc implementation. Called by ``pyfunc.load_pyfunc``.
+
+    :param path: Local filesystem path to the MLflow Model with the ``gluon`` flavor.
+    """
+    m = load_model(path, mx.current_context())
+    return _GluonModelWrapper(m)
+
+
+@experimental
+def save_model(gluon_model, path, mlflow_model=Model()):
```
Comment (dbczumar) [inline mlflow/gluon.py]: @cosmincatalin Can you add a `conda_env` parameter to the `save_model` method as well as a default conda environment that contains all of the dependencies necessary for using the gluon flavor? For an example, see the `get_default_conda_env()` method in the `mlflow.sklearn` module: https://github.com/mlflow/mlflow/blob/master/mlflow/sklearn.py#L39-L55.

This will ensure that all relevant dependencies are installed when users to serve or deploy a gluon model (e.g., via the `mlflow models serve` CLI).

**Review by dbczumar — CHANGES_REQUESTED** [review]
@cosmincatalin I've restarted the errant test job. I've left a couple comments regarding the need for a conda environment for use with the pyfunc flavor. Is it possible to add a unit test for loading the Gluon model as a python function (`pyfunc`) and evaluating it on example data?

File: mlflow/gluon.py (hunk 11 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: Done

File: mlflow/gluon.py (hunk 12 above)
Comment (cosmincatalin) [reply mlflow/gluon.py]: Done

**Comment by cosmincatalin** [comment]
One of the tests in Travis still breaks unfortunately. I'm not sure what the problem is with it.

**Comment by cosmincatalin** [comment]
I've managed to fix everything. 😄

**Comment by dbczumar** [comment]
@cosmincatalin I pushed a fix to your branch to pass the conda environment to the pyfunc configuration when saving the model. Additionally, I added an `MXNet Gluon` entry to the MLflow Models docs. Please let me know if this looks okay.

Before merging, I'd like to add tests to ensure that conda environments are persisted correctly and added to the pyfunc configuration. I'll aim to do this within the next 24 hours, but please feel free to do so in the mean time!

**Comment by cosmincatalin** [comment]
@dbczumar Everything looks fine with me in the documentation. Please go ahead and add the tests.

**Review by dbczumar — APPROVED** [review]
@cosmincatalin I pushed the remaining tests to the branch, as well as tests for scoring. I also updated the `pyfunc` representation of Gluon models to conform to the `pyfunc` prediction interface (https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#inference-api). Please let me know if you have any questions about that interface. With these changes, this PR looks great and should be ready to merge once tests pass. Thank you for your contribution!

**Comment by cosmincatalin** [comment]
Everything looks good on my side. I can't wait to see it released. I'll prepare an example scenario on my blog as well.

**Comment by dbczumar** [comment]
@cosmincatalin Awesome! Just merged. Looking forward to your blog post!

**Comment by cosmincatalin** [comment]
@dbczumar Some half a year later I have the blog post if you're interested in advertising it on some channels: https://cosminsanda.com/posts/experiment-tracking-with-mlflow-inside-amazon-sagemaker/
